go/logger/buffer.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2019 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package logger
+
+import (
+	"bufio"
+	"io"
+	"sync"
+	"time"
+)
+
+type autoFlushingBufferedWriter struct {
+	lock           sync.RWMutex
+	bufferedWriter *bufio.Writer
+	backupWriter   *bufio.Writer
+
+	frequency        time.Duration
+	timer            *time.Timer
+	currentlyWaiting bool
+	shutdown         chan struct{}
+}
+
+var _ io.Writer = &autoFlushingBufferedWriter{}
+
+func (writer *autoFlushingBufferedWriter) backgroundFlush() {
+	for {
+		select {
+		case <-writer.timer.C:
+			// Swap out active and backup writers
+			writer.lock.Lock()
+			writer.currentlyWaiting = false
+			previouslyActive := writer.bufferedWriter
+			writer.bufferedWriter = writer.backupWriter
+			writer.lock.Unlock()
+
+			// Flush the previously active writer and complete the swap. We do
+			// not need to complete the swap under lock because this is the
+			// only goroutine accessing backupWriter.
+			writer.backupWriter = previouslyActive
+			writer.backupWriter.Flush()
+		case <-writer.shutdown:
+			writer.timer.Stop()
+			writer.bufferedWriter.Flush()
+			return
+		}
+	}
+}
+
+// NewAutoFlushingBufferedWriter returns an io.Writer that buffers its output
+// and flushes automatically after `flushFrequency`.
+func NewAutoFlushingBufferedWriter(baseWriter io.Writer,
+	flushFrequency time.Duration) (w io.Writer, shutdown chan struct{}) {
+	result := &autoFlushingBufferedWriter{
+		bufferedWriter: bufio.NewWriter(baseWriter),
+		backupWriter:   bufio.NewWriter(baseWriter),
+		frequency:      flushFrequency,
+		timer:          time.NewTimer(flushFrequency),
+	}
+	go result.backgroundFlush()
+	return result, result.shutdown
+}
+
+func (writer *autoFlushingBufferedWriter) Write(p []byte) (int, error) {
+	writer.lock.RLock()
```

Comment:
Let's add a comment explaining that we're using an `RLock` on a "write" because the locked resource is only experiencing a read here.